File: typescript/examples/getting-started/03a-tools-call.ts
// Tools: non-streaming call with tool definitions
// Defines a tool and gets the result including any tool_calls the LLM wants to make.
import { z } from "zod";
import { Opper } from "../../src/index.js";

const opper = new Opper();

const result = await opper.call("sdk-test-tool-use", {
  input_schema: z.object({
    question: z.string().describe("The user's question"),
  }),
  output_schema: z.object({
    answer: z.string().optional().describe("The assistant's text response"),
    tool_calls: z.array(z.object({
      name: z.string(),
      arguments: z.record(z.string(), z.unknown()),
    })).optional().describe("Tool calls requested by the model"),
  }),
  input: { question: "What is the current weather in Stockholm?" },
  model: "anthropic/claude-sonnet-4.6",
  tools: [
    {
      name: "get_weather",
      description: "Get the current weather for a city",
      parameters: z.object({
        city: z.string().describe("City name"),
        unit: z.enum(["celsius", "fahrenheit"]).optional().describe("Temperature unit"),
      }),
    },
  ],
});

console.log("Answer:", result.data.answer);
console.log("Tool calls:", JSON.stringify(result.data.tool_calls, null, 2));

File: python/examples/getting-started/03a_tools_call.py
# Tools: non-streaming call with tool definitions
# Schemas can be Pydantic models, raw dicts, or Python types — mix and match freely.
from pydantic import BaseModel, Field

from opperai import Opper

opper = Opper()


# ── Pydantic for structured schemas (optional — raw dicts work too) ─────────


class AssistantResponse(BaseModel):
    answer: str | None = Field(default=None, description="The assistant's text response")
    tool_calls: list[dict] | None = Field(default=None, description="Tool calls requested by the model")


class WeatherParams(BaseModel):
    city: str = Field(description="City name")
    unit: str | None = Field(default=None, description="Temperature unit: celsius or fahrenheit")


# ── Call with tools ─────────────────────────────────────────────────────────

result = opper.call(
    "sdk-test-tool-use",
    input={"question": "What is the current weather in Stockholm?"},  # plain dict
    input_schema={"type": "object", "properties": {"question": {"type": "string"}}},  # raw JSON Schema
    output_schema=AssistantResponse,  # Pydantic model
    model="anthropic/claude-sonnet-4.6",
    tools=[
        {
            "name": "get_weather",
            "description": "Get the current weather for a city",
            "parameters": WeatherParams,  # Pydantic model for tool params
        },
    ],
)

# When output_schema is a Pydantic model, result.data is typed
print("Answer:", result.data.answer)
print("Tool calls:", result.data.tool_calls)
